<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Откройте-и-изучите-файл" data-toc-modified-id="Откройте-и-изучите-файл-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Откройте и изучите файл</a></span></li><li><span><a href="#Разбейте-данные-на-выборки" data-toc-modified-id="Разбейте-данные-на-выборки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Разбейте данные на выборки</a></span></li><li><span><a href="#Исследуйте-модели" data-toc-modified-id="Исследуйте-модели-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Исследуйте модели</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Итоги-исследования" data-toc-modified-id="Итоги-исследования-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Итоги исследования</a></span></li></ul></li><li><span><a href="#Проверьте-модель-на-тестовой-выборке" data-toc-modified-id="Проверьте-модель-на-тестовой-выборке-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Проверьте модель на тестовой выборке</a></span><ul class="toc-item"><li><span><a href="#Тест-лучшего-дерева-решений" data-toc-modified-id="Тест-лучшего-дерева-решений-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Тест лучшего дерева решений</a></span></li><li><span><a href="#Тест-лучшего-случайного-леса" data-toc-modified-id="Тест-лучшего-случайного-леса-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Тест лучшего случайного леса</a></span></li><li><span><a href="#Тест-логистической-регрессии" data-toc-modified-id="Тест-логистической-регрессии-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Тест логистической регрессии</a></span></li><li><span><a href="#Итоги-теста-лучших-моделей" data-toc-modified-id="Итоги-теста-лучших-моделей-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Итоги теста лучших моделей</a></span></li></ul></li><li><span><a href="#(бонус)-Проверьте-модели-на-адекватность" data-toc-modified-id="(бонус)-Проверьте-модели-на-адекватность-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>(бонус) Проверьте модели на адекватность</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Рекомендация тарифов

В нашем распоряжении данные о поведении клиентов, которые уже перешли на тарифы "Смарт" и "Ультра" оператора мобильной связи "Мегалайн". На их основе нужно построить модель для задачи классификации, которая выберет подходящий конкретному пользователю тариф. Известно, что предобработка данных не понадобится — её уже сделали.

Доля правильных ответов построенной модели должна быть не менее 75%

## Описание данных

Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:

- `сalls` — количество звонков,
- `minutes` — суммарная длительность звонков в минутах,
- `messages` — количество sms-сообщений,
- `mb_used` — израсходованный интернет-трафик в Мб,
- `is_ultra` — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

## Импорт библиотек

Прежде, чем приступить к изучению данных, импортируем необходимые библиотеки

In [15]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

## Откройте и изучите файл

Так как проект выполняется локально, используем конструкцию `try - except`, чтобы при загрузке файла на ревью, не возникло проблем с путём к файлу, а также применим метод `info`, чтобы убедиться в отсуствии пропусков, корректности типов данных и визуально взглянем на 10 случайных строк таблицы:

In [16]:
try:
    users = pd.read_csv('/datasets/users_behavior.csv')
except:
    users = pd.read_csv('users_behavior.csv')
    
display(users.sample(10))

calls  minutes  messages   mb_used  is_ultra
2798   71.0   485.12      46.0  16702.03         0
2482   53.0   370.88      82.0  21276.92         1
2106   70.0   453.96      52.0  17087.02         0
2906   52.0   346.55      19.0  16459.11         0
2390    3.0    48.74       2.0    439.62         1
1771   53.0   351.50      31.0   9766.09         0
2310   93.0   648.70       0.0  12458.34         1
1002   35.0   197.96      45.0  21425.84         0
545    60.0   466.61      67.0  18507.05         0
1869   36.0   268.66      13.0   8209.15         0

Визуально всё выглядит корректно

In [17]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


Пропусков не обнаружено, типы вполне себе подходящие, названия столбцов в порядке.

## Разбейте данные на выборки

Нам необходимо три выборки. Обучающая, валидационная и тестовая. Объёмы выборок в процентном соотношении должны быть соответственно 60%, 20% и 20%. Сделаем это последовательно:

In [18]:
# создаём датасет со свойствами
user_features = users.drop(['is_ultra'], axis=1)
# создаём датасет с целевыми признаками
tariff_target = users['is_ultra']

# создаём обучающий датасет объёмом 60% от исходного
features_train, features_valid, target_train, target_valid = train_test_split(
    user_features, tariff_target, test_size=0.4, random_state=12345)
# оставшуюся часть делим пополам и получаем части по 20% от исходного объёма
features_valid, features_test, target_valid, target_test = train_test_split(
    features_valid, target_valid, test_size=0.5, random_state=12345)

# проверим, как резделены данные по объёму
for data in [features_train, features_valid, features_test, target_train, target_valid, target_test]:
    print(f'Размерность: {data.shape}')    

Размерность: (1928, 4)
Размерность: (643, 4)
Размерность: (643, 4)
Размерность: (1928,)
Размерность: (643,)
Размерность: (643,)


Видно, что целевые соотношения объёмов выборок получены - 1928 объектов в обучающей, и по 643 в валидационной и тестовой выборках. Проверим визуально первые две строки выборок:

In [5]:
for data in [features_train, features_valid, features_test]:
    display(data.head(2))

calls  minutes  messages   mb_used
3027   60.0   431.56      26.0  14751.26
434    33.0   265.17      59.0  17398.02

calls  minutes  messages   mb_used
1386   92.0   536.96      18.0  20193.90
3124   40.0   286.57      17.0  17918.75

calls  minutes  messages   mb_used
160    61.0   495.11       8.0  10891.23
2498   80.0   555.04      28.0  28083.58

Сравнивая первые строки получившихся выборок делаем вывод, что они разные.

## Исследуйте модели

Исследование моделей будем проводить с помощью цикла, перебирающего значения гиперпараметров, выводя показатели точности каждой итерации.

### Дерево решений

Для дерева решений перебираем гиперпараметр "максимальная глубина":

In [6]:
# создаем переменную для модели с лучшим показателем точности
best_tree_model = None
# создаем переменную для лучшего показателя точности
best_tree_accuracy = 0
# создаем переменную для значения максимальной глубины, обеспечившей максимальную точность
best_tree_depth = 0

# пробежимся по значениям глубины от 1 до 10
for depth in range(1, 11):
    model_tree = DecisionTreeClassifier(max_depth=depth, random_state=12345) # создаём дерево
    model_tree.fit(features_train, target_train) # тренируем на тренировочной выборке
    prediction_tree_valid = model_tree.predict(features_valid) # предсказываем результаты по валидационным свойствам
    accuracy_tree_valid = accuracy_score(target_valid, prediction_tree_valid) # считаем точность
    print(f'Точность дерева решений при максимальной глубине {depth}: {accuracy_tree_valid}')
    # условной конструкцией записываем в ранее созданные переменные лучшие показатели
    if accuracy_tree_valid > best_tree_accuracy:
        best_tree_model = model_tree
        best_tree_accuracy = accuracy_tree_valid
        best_tree_depth = depth

print() # отделим лучший результат от промежуточных
print('Лучшая модель дерева решений:')
print(f'- Глубина: {best_tree_depth}')
print(f'- Точность: {(best_tree_accuracy * 100).round(2)}%')

Точность дерева решений при максимальной глубине 1: 0.7542768273716952
Точность дерева решений при максимальной глубине 2: 0.7822706065318819
Точность дерева решений при максимальной глубине 3: 0.7853810264385692
Точность дерева решений при максимальной глубине 4: 0.7791601866251944
Точность дерева решений при максимальной глубине 5: 0.7791601866251944
Точность дерева решений при максимальной глубине 6: 0.7838258164852255
Точность дерева решений при максимальной глубине 7: 0.7822706065318819
Точность дерева решений при максимальной глубине 8: 0.7791601866251944
Точность дерева решений при максимальной глубине 9: 0.7822706065318819
Точность дерева решений при максимальной глубине 10: 0.7744945567651633

Лучшая модель дерева решений:
- Глубина: 3
- Точность: 78.54%


Видим, что минимальное значение точности оказалось при глубине 1 - 75%. При остальных значениях - 77-78%. Лучшая модель с глубиной 3 показала точность 78,54%

### Случайный лес

Для случайного леса аналогичным способом перебираем максимальную глубину и количество деревьев в модели.

In [7]:
# создаём переменные для хранения лучших показателей
best_forest_model = None
best_forest_accuracy = 0
best_forest_est = 0
best_forest_depth = 0

# циклом перебираем количество деревьев от 10 до 50 включительно с шагом 10, глубину так же - от 1 до 10
for est in range(10, 51, 10):
    for depth in range(1, 11):
        model_forest = RandomForestClassifier(max_depth=depth, n_estimators=est, random_state=12345)
        model_forest.fit(features_train, target_train)
        accuracy_forest_valid = model_forest.score(features_valid, target_valid)
        print(f'Точность случайного леса из {est} деревьев при максимальной глубине {depth}: {accuracy_forest_valid}')
        if accuracy_forest_valid > best_forest_accuracy:
            best_forest_model = model_forest
            best_forest_accuracy = accuracy_forest_valid
            best_forest_depth = depth
            best_forest_est = est

print() # отделим лучший результат от промежуточных
print('Лучшая модель случайного леса:')
print(f'- Количество деревьев: {best_forest_est}')
print(f'- Глубина: {best_forest_depth}')
print(f'- Точность: {(best_forest_accuracy * 100).round(2)}%')

Точность случайного леса из 10 деревьев при максимальной глубине 1: 0.7558320373250389
Точность случайного леса из 10 деревьев при максимальной глубине 2: 0.7776049766718507
Точность случайного леса из 10 деревьев при максимальной глубине 3: 0.7853810264385692
Точность случайного леса из 10 деревьев при максимальной глубине 4: 0.7900466562986003
Точность случайного леса из 10 деревьев при максимальной глубине 5: 0.7931570762052877
Точность случайного леса из 10 деревьев при максимальной глубине 6: 0.8009331259720062
Точность случайного леса из 10 деревьев при максимальной глубине 7: 0.7947122861586314
Точность случайного леса из 10 деревьев при максимальной глубине 8: 0.7962674961119751
Точность случайного леса из 10 деревьев при максимальной глубине 9: 0.7853810264385692
Точность случайного леса из 10 деревьев при максимальной глубине 10: 0.7916018662519441
Точность случайного леса из 20 деревьев при максимальной глубине 1: 0.7667185069984448
Точность случайного леса из 20 деревьев пр

Видно, что количество деревьев в выбранном диапазоне не особо влияет на точность исследования - максимальная точность варьируется находится в диапазоне 80-81% - прирост по сравнению со случайным деревом в 1,5 - 2%. Скорее сам факт использования модели случайного леса даёт увеличение точности.

### Логистическая регрессия

Для логистической регрессии аналогично перебираем количество итераций:

In [8]:
best_logistic_model = None
best_logistic_accuracy = 0
best_logistic_iter = 0

# перебираем от 100 до 1000 с шагом 100
for ite in range(100, 1000, 100):
    model_logistic = LogisticRegression(max_iter = ite, solver='lbfgs', random_state=12345)
    model_logistic.fit(features_train, target_train)
    accuracy_logistic_valid = model_logistic.score(features_valid, target_valid)
    print(f'Точность логистической регрессии при {ite} итерациях: {accuracy_logistic_valid}')
    
    if accuracy_logistic_valid > best_logistic_accuracy:
        best_logistic_model = model_logistic
        best_logistic_accuracy = accuracy_logistic_valid
        best_logistic_iter = ite

print() # отделим лучший результат от промежуточных
print('Лучшая модель логистической регрессии:')
print(f'- Количество итераций: {best_logistic_iter}')
print(f'- Точность: {(best_logistic_accuracy * 100).round(2)}%')

Точность логистической регрессии при 100 итерациях: 0.7107309486780715
Точность логистической регрессии при 200 итерациях: 0.7107309486780715
Точность логистической регрессии при 300 итерациях: 0.7107309486780715
Точность логистической регрессии при 400 итерациях: 0.7107309486780715
Точность логистической регрессии при 500 итерациях: 0.7107309486780715
Точность логистической регрессии при 600 итерациях: 0.7107309486780715
Точность логистической регрессии при 700 итерациях: 0.7107309486780715
Точность логистической регрессии при 800 итерациях: 0.7107309486780715
Точность логистической регрессии при 900 итерациях: 0.7107309486780715

Лучшая модель логистической регрессии:
- Количество итераций: 100
- Точность: 71.07%


Никакой зависимости точности от количества итераций не видно... Предполагаю, что это максимально возможная точность данной модели.

### Итоги исследования

- Максимальную точность на валидационной выборке показала модель случайного леса - 80,87%
- Минимальная точность у логистической регрессии - 71,07%
- Дерево решений по точности ближе к случайному лесу, поэтому думаю, что для бинарной классификации модель логистической регрессии не подходит.

## Проверьте модель на тестовой выборке

Проверим модели на тестовой выборке, заодно сравним точности предсказаний по всем выборкам (и обучающей тоже):

### Тест лучшего дерева решений

In [9]:
# посчитаем результат лучшей модели дерева решений на обучающей выборке
train_tree_prediction = best_tree_model.predict(features_train)
train_tree_accuracy = accuracy_score(target_train, train_tree_prediction)

# применяем модель и считаем точность на тестовой выборке
best_tree_prediction = best_tree_model.predict(features_test)
test_tree_accuracy = accuracy_score(target_test, best_tree_prediction)

print(f'Точность предсказаний решающего дерева по тестовой выборке: {(test_tree_accuracy * 100).round(2)}%')
print(f'Точность предсказаний решающего дерева по валидационой выборке: {(best_tree_accuracy * 100).round(2)}%')
print(f'Точность предсказаний решающего дерева по обучающей выборке: {(train_tree_accuracy * 100).round(2)}%')

Точность предсказаний решающего дерева по тестовой выборке: 77.92%
Точность предсказаний решающего дерева по валидационой выборке: 78.54%
Точность предсказаний решающего дерева по обучающей выборке: 80.76%


Судя по результатам на тестовой и обучающей выборках, модель очень даже обучена.

### Тест лучшего случайного леса

In [10]:
# посчитаем результат лучшей модели случайного леса на обучающей выборке
train_forest_accuracy = best_forest_model.score(features_train, target_train)

# применяем модель и считаем точность на тестовой выборке
test_forest_accuracy = best_forest_model.score(features_test, target_test)

print(f'Точность предсказаний случайного леса по тестовой выборке: {(test_forest_accuracy * 100).round(2)}%')
print(f'Точность предсказаний случайного леса по валидационой выборке: {(best_forest_accuracy * 100).round(2)}%')
print(f'Точность предсказаний решающего дерева по обучающей выборке: {(train_forest_accuracy * 100).round(2)}%')

Точность предсказаний случайного леса по тестовой выборке: 79.63%
Точность предсказаний случайного леса по валидационой выборке: 80.87%
Точность предсказаний решающего дерева по обучающей выборке: 87.5%


Разрыв в точности между обучающей и тестовой выборкой приличный, но результат на тесте показан достаточно высокий. Модель слегка переобучена.

### Тест логистической регрессии

In [11]:
# посчитаем результат лучшей модели на обучающей выборке
train_logistic_accuracy = best_logistic_model.score(features_train, target_train)
# применяем модель и считаем точность на тестовой выборке
test_logistic_accuracy = best_logistic_model.score(features_test, target_test)

print(f'Точность предсказаний логистической регрессии по тестовой выборке: {(test_logistic_accuracy * 100).round(2)}%')
print(f'Точность предсказаний логистической регрессии по валидационой выборке: {(best_logistic_accuracy * 100).round(2)}%')
print(f'Точность предсказаний решающего дерева по обучающей выборке: {(train_logistic_accuracy * 100).round(2)}%')

Точность предсказаний логистической регрессии по тестовой выборке: 68.43%
Точность предсказаний логистической регрессии по валидационой выборке: 71.07%
Точность предсказаний решающего дерева по обучающей выборке: 71.32%


Результат был низкий, а стал ещё хуже. Низкие результаты и по тестам и в обучении - модель недообучена (или необучаема в данных условиях?)

### Итоги теста лучших моделей

- Лучшая модель - случайный лес: 79,63% точности. При этом модель переобучена
- Худшая модель - логистическая регрессия: 68,43% точности. Модель недообучена.
- Все модели на тестовой выборке показали результат хуже, чем на валидационной:
    - Дерево решений - на полпроцента
    - Случайный лес - на процент
    - Логистическая регрессия - на два с половиной процента

## (бонус) Проверьте модели на адекватность

Судя по теоретическим материалам, проверить модель на адекватность нужно сравнить результаты её применения со случайным выбором. Так как мы имеем дело с бинарной классификацией, нужно понять, какова вероятность случайно попасть в тариф "Ультра". И простым вычитанием из единицы мы получим вероятность попасть в тариф "Смарт".

In [12]:
# считаем вероятность случайно выбрать тариф "Ультра"
ultra_prob = target_test.sum() / target_test.count()
# считаем вероятность случайно выбрать тариф "Смарт"
smart_prob = 1 - ultra_prob
print(f'Вероятность попасть в "Ультра" в тестовой выборке: {(ultra_prob * 100).round(2)}%')
print(f'Вероятность попасть в "Смарт" в тестовой выборке: {(smart_prob * 100).round(2)}%')

Вероятность попасть в "Ультра" в тестовой выборке: 31.57%
Вероятность попасть в "Смарт" в тестовой выборке: 68.43%


Таким образом мы понимаем, что если мы при определении тарифов будем **всегда** выбирать "Смарт", то будем правы в 68,43% случаев. Это значение и будет границей адекватности. Точность нашей модели должна быть ощутимо выше, чтобы было целесообразно её применять. Проверим наши модели на адекватность и сделаем выводы в автоматическом режиме (программисты мы или где? :-)):

In [13]:
# создадим словарь с ключами-моделями и значениями-точностями в тесте
models_dict = {
    'Логистическая регрессия' : test_logistic_accuracy ,
    'Случайный лес'  : test_forest_accuracy,
    'Дерево решений' : test_tree_accuracy}

# пробежимся циклом по словарю, сравнивая точность модели в тесте со случайным выбором.
for model, accuracy in models_dict.items():
    if smart_prob == accuracy:
        print(f'{model} жульничает!')
    elif smart_prob > accuracy:
        print(f'{model} вредит!')
    else:
        print(f'{model} круче случайного выбора на {((accuracy - smart_prob) * 100).round(2)}%')

Логистическая регрессия жульничает!
Случайный лес круче случайного выбора на 11.2%
Дерево решений круче случайного выбора на 9.49%


Точность предсказаний логистической регрессии в точности равна вероятности попасть в тариф "Смарт" - **68,43%**. Вывод: **регрессия - жулик**! Проставила везде "Смарт". Модель явно неадекватна.

А вот дерево решений и случайный лес показали себя лучше, и мы вполне можем их применять.

In [14]:
# код ревьюера

logistic_pred = best_logistic_model.predict(features_test)
logistic_pred.mean()

0.024883359253499222

## Общий вывод

Получив предобработанные данные и убедившись в том, что они действительно предобработаны, мы последовательно разбили исходный датасет на три части: **60% - обучающая выборка, 20% - валидационная выборка, 20% - тестовая выборка**.

По результатам **исследования**:
- лучшая модель по точности - **случайный лес - 80,87%**. Сама модель даёт прирост в точности на 1,5 - 2%, количество деревьев и глубина в исследуемых диапазонах гиперпараметров такого прироста не дают.
- Минимальная точность у **логистической регрессии - 71,07%**, для бинарной классификации модель логистической регрессии не очень подходит.
- **Дерево решений** по точности ближе к случайному лесу - **78,54%**

На этапе **тестирования** моделей увидели, что все модели показали результаты хуже, чем на валидации:
- Дерево решений - на полпроцента
- Случайный лес - на процент
- Логистическая регрессия - на два с половиной процента

Тем не менее по точности распределение мест не поменялось:
- Лучшая модель - **случайный лес: 79,63%** точности. Модель переобучена
- Худшая модель - **логистическая регрессия: 68,43%** точности. Модель недообучена.
- **Дерево решений - 77,92%** точности. Надёжный, крепкий середняк.

**Проверка на адекватность** выявила, что **логистическая регрессия жулик** - её точность точно такая же как вероятность случайно выбрать тариф "Смарт". **Не выбирала**, а просто всем проставила нули.

А растения показали себя так:
- Случайный лес круче случайного выбора на 11.2%
- Дерево решений круче случайного выбора на 9.49%

**Для задачи классификации тарифов рекомендую применять модель `случайного леса` с долей правильных ответов на тесте `79,63%`**

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
